ndt_slam:
  scan_publisher_topic: "/normal_distribution"        # visualizer topi of ndt scan
  map_publisher_topic: "/aligned_normal_distribution" # visualizer topic of ndt submap
  scan_subscriber_topic: "/Navtech/PointCloud"        # pcl2 topic
  imu_subscriber_topic: "/imu/transformed/data"       # imu topic in base_frame
  odom_publisher_topic: "/ndt_odom"                   # odometry publisher topic 
  fixed_frame: "world"                                # map frame
  odom_frame: "navtech"                               # odometry frame
  base_frame: "navtech"                               # base frame, should be accurate for motion model
  sensor_frame: "navtech"                             # sensor frame
  visualize_ogm: false                                # visualize ogm
  visualize_current_scan: false                       # visualize ndt of scam
  visualize_current_map: false                        # visualize ndt submap
  visualize_path: true                                # visualize pose graph
  use_imu: false                                      # if use imu 
  initialize_from_tf: false                           # if true, transform odom->base_link has to be available
  visualizer_frequency: 0.5                           # trigger visualizer with this frequency
  pose_graph_frequency: 0.2                           # trigger pose graph optimization with this frequency
  search_loop_closure_frequency: 0.8                  # trigger loop closure search with this frequency
  online: false
  rosbag_path: "/root/catkin_ws/rosbags/oxford_raw/oxford-eval-sequences/2019-01-10-12-32-52-radar-oxford-10k/radar/2019-01-10-12-32-52-radar-oxford-10k.bag" 
  #rosbag_path: "/root/catkin_ws/rosbags/oxford_raw/oxford-eval-sequences/2019-01-16-13-09-37-radar-oxford-10k/radar/2019-01-16-13-09-37-radar-oxford-10k.bag" 
  #rosbag_path: "/root/catkin_ws/rosbags/oxford_raw/oxford-eval-sequences/2019-01-17-13-26-39-radar-oxford-10k/radar/2019-01-17-13-26-39-radar-oxford-10k.bag"
  #rosbag_path: "/root/catkin_ws/rosbags/oxford_raw/oxford-eval-sequences/2019-01-18-14-14-42-radar-oxford-10k/radar/2019-01-18-14-14-42-radar-oxford-10k.bag" 
  #rosbag_path: "/root/catkin_ws/rosbags/oxford_raw/oxford-eval-sequences/2019-01-18-15-20-12-radar-oxford-10k/radar/2019-01-18-15-20-12-radar-oxford-10k.bag" 
  #rosbag_path: "/root/catkin_ws/rosbags/oxford_raw/oxford-eval-sequences/2019-01-10-11-46-21-radar-oxford-10k/radar/2019-01-10-11-46-21-radar-oxford-10k.bag"  
  #rosbag_path: "/root/catkin_ws/rosbags/oxford_raw/oxford-eval-sequences/2019-01-16-11-53-11-radar-oxford-10k/radar/2019-01-16-11-53-11-radar-oxford-10k.bag" 
  #rosbag_path: "/root/catkin_ws/rosbags/oxford_raw/oxford-eval-sequences/2019-01-18-14-46-59-radar-oxford-10k/radar/2019-01-18-14-46-59-radar-oxford-10k.bag" 

global_fuser:                                         
  loss_function_scale: 750                # scale of robust loss in pose graph optimization. Not needed if use_robust_loss is false
  use_robust_loss: false                  # if true, use Huber Loss in pose graph optimization

local_fuser:
  use_scan_context_as_loop_closure: true         # use this
  compute_dfs_loop_closure: false                # you can use this matcher instead, but is not recommended
  max_data_association_mahalanobis_dist: 0.5     # for csm, considere canditate node with maximum mahalanobis distance
  
  loop_sqrtI: [1.0, 0.0, 0.0,                    # d_x, d_y, d_th
               0.0, 1.0, 0.0,
               0.0, 0.0, 0.1]

  insertion_step: 2                        # every n scan is considered a keyframe
  loop_closure_gnc_steps: 10               # graduation steps in loop closure refinement (lcr). If omitted, set to same value as in local motion estimation
  loop_closure_max_cs_divergence: 4.5      # cs divergence threshold
  loop_closure_scale: 0.5                  # scale of robust loss function in lcr. If omitted, set to same value as in local motion estimation
  loop_closure_weight: 1                   # weight of loop closure constraint in pose graph optimization
  submap_size_poses: 20                    # number of states in each submap
  submap_overlap: 10                       # for this number of states, also the previous submap is used in matchin
  use_scan_context_as_loop_closure: true   # select scd for loop closure. Should always be true

ndt_map:
  max_neighbor_linf_distance: 10.0         # size of search window for neares neighbors
  min_points_per_cell: 10                  # minimal number of points to generate distribution in cell
  resolution: 3.5                          # side length of an NDT cell
  size_x: 400                              # size of underlying map in cells
  size_y: 400 

ndt_matcher:
  gnc_control_parameter_divisor: 1.1       # k_mu to reduce convexity in each graduation step
  gnc_steps: 2                             # number of graduation steps
  lookup_distribution: true                # if true, use l2 distance between distributions. Else, l2 distance of the grid cells holding the distributions
  loss_function_convexity: -2.0            # convexity of the robust loss function
  loss_function_scale: 1                   # scale of the robust loss function
  n_results_nn_lookup: 2                   # number of neighbors for each moving ndt cell
  optimize_on_manifold: true               # if true, optimize over Lie group element, else over three vector representation
  smoothing_steps: 3                       # lag size
  use_constant_velocity_model: true        # if true, use constant velocity, else constant acceleration
  use_intensity_as_dimension: true         # if true, use intensity in matching

  ndt_weight: 5000.0                       # weight of ndt matching in local motion estimation
  covariance_scaling_factor: 0.01          # weight of the motion model. The individual covariance matrix is given in the overall config
  weight_imu: 64                           # weight of imu rotation estimate
  weight_imu_bias: 750000.1                # weight if imu bias

  motion_sqrtI: [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,      # x, y, theta (submap frame), v_x, v_y, v_th, a_x, a_y (robot frame)
                 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                 0.0, 0.0, 10, 0.0, 0.0, 0.0, 0.0, 0.0,
                 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0,
                 0.0, 0.0, 0.0, 0.0, 3.0, 0.0, 0.0, 0.0,
                 0.0, 0.0, 0.0, 0.0, 0.0, 0.1, 0.0, 0.0,
                 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 20.0, 0.0,
                 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 60.0]
  initial_imu_bias: 0.0                            # initial value of imu bias
  max_iteration: 200                               # in each graduated optimization step, limit number of iterations
  pose_reject_translation: 5                       # if the estimate deviates more than this from intial guess, the initial guess is further considered
  pose_reject_rotation: 2
  use_analytic_expressions_for_optimization: false # only available if not using manifold

  # use of csm by default deactivated and not recommended
  csm_cost_threshold: 0.82           # if this cost is exceeded, do not evaluate pose candidate further
  csm_linear_step: 0.4               # linear space between candidates in finest level
  csm_max_px_accurate_range: 4.0     # angular resolution is defined based on this range
  csm_n_iter: 2                      # number of levels
  csm_window_angular: 0.45           # maximum angular search window
  csm_window_linear: 4.5             # maximum linear search window

radar_preprocessor:
  min_range: 2.0                          # minimum range of sensor. Should filter out radome noise
  max_range: 100                          # maximum range of sensor
  min_intensity: 70                       # minimal intensity. Should filter out speckle noise  
  beam_distance_increment_threshold: 0.12 # connectivity criterion for beam-wise clustering. Sensor-dependent                                                                


scan_context:                      # no need to tune, work pretty well out of the box
  assumed_drift: 0.05              # odometric drift
  dist_thresh: 0.7                 # maximum combined descriptor and odometry distance
  intensity_factor: 0.01           # multiplier for intensity values
  max_radius: 90.0                 # max range of descriptor
  num_candidates_from_tree: 10     # in first lookup stage, retrieve this number of candidate
  num_exclude_recent: 100          # the last n keyframes are not considered as keyframe
  num_ring: 30                     # number of range bins
  num_sector: 120                  # number of azimuth bins 
  odom_eps: 4                      # allowed linear displacement to consider same place
  odom_weight: 0.05                # weight of odometry similarity in combined distance
  search_ratio: 1                  # only search ratio of keyframes
  tree_making_period: 10           # rebuild tree after inserting n keyframes

ogm_map:
  size_x: 90 # size of underlying map in m
  size_y: 40 
  resolution: 0.1 # side length of OGM cell

ndt_cell:
  beam_cov: [0.0349208,     0, 0, 
                     0, 0.001, 0, 
                     0,     0, 10]
  use_pndt: false # hard to tune, better not touch it
